src/utils/havoc.js

```diff
@@ -173,57 +166,26 @@ class ObjectValueHavocingVisitor {
 
     // if this object wasn't already havoced, we need mark it as havoced
     // so that any mutation and property access get tracked after this.
-    if (obj.mightNotBeHavocedObject()) {
-      obj.havoc();
-      if (obj.symbols.size > 0) {
-        throw new FatalError("TODO: Support havocing objects with symbols");
-      }
-      if (obj.unknownProperty !== undefined) {
-        // TODO: Support unknown properties, or throw FatalError.
-        // We have repros, e.g. test/serializer/additional-functions/ArrayConcat.js.
-      }
-      // TODO: We should emit current value and then reset value for all *internal slots*; this will require deep serializer support; or throw FatalError when we detect any non-initial values in internal slots.
-      let realmGenerator = this.realm.generator;
-      for (let [name, propertyBinding] of obj.properties) {
-        // ignore properties with their correct default values
-        if (this.getHeapInspector().canIgnoreProperty(obj, name)) continue;
-
-        let descriptor = propertyBinding.descriptor;
-        if (descriptor === undefined) {
-          // TODO: This happens, e.g. test/serializer/pure-functions/ObjectAssign2.js
-          // If it indeed means deleted binding, should we initialize descriptor with a deleted value?
-          if (realmGenerator !== undefined) realmGenerator.emitPropertyDelete(obj, name);
-        } else {
-          let value = descriptor.value;
-          invariant(
-            value === undefined || value instanceof Value,
-            "cannot be an array because we are not dealing with intrinsics here"
-          );
-          if (value === undefined) {
-            // TODO: Deal with accessor properties
-            // We have repros, e.g. test/serializer/pure-functions/AbstractPropertyObjectKeyAssignment.js
-          } else {
-            invariant(value instanceof Value);
-            if (value instanceof EmptyValue) {
-              if (realmGenerator !== undefined) realmGenerator.emitPropertyDelete(obj, name);
-            } else {
-              if (realmGenerator !== undefined) {
-                let targetDescriptor = this.getHeapInspector().getTargetIntegrityDescriptor(obj);
-                if (!isReactElement(obj)) {
-                  if (
-                    descriptor.writable !== targetDescriptor.writable ||
-                    descriptor.configurable !== targetDescriptor.configurable
-                  ) {
-                    realmGenerator.emitDefineProperty(obj, name, descriptor);
-                  } else {
-                    realmGenerator.emitPropertyAssignment(obj, name, value);
-                  }
-                }
-              }
-            }
-          }
+    if (!obj.isLeakedObject()) {
```

Comment:
I'm uncomfortable with this. Havocing an object is not the same as leaking it. This conflates the two concepts.

Reply:
This comment provides no information, beyond your being uncomfortable with this. The point of the renames I did was to clarify the difference in the code between these two concepts. Do you disagree with the changed use at this code site? Do you disagree with it in general? Please be more precise.

Reply:
A routine that is named "havoc" ends up "leaking" an object. That smells wrong.

Reply:
This entire file / class should get renamed away from "havoc" to "leaking". Materialization and havocing are side effects of leaking.

Reply:
If we come across a use case where we need to havoc something without leaking it, this approach will bite us. Right now, I'm not convinced that there is such a use case, so this may turn out to be OK. It just smells a bit off to me.